% automatic manuscript creation for nondetects
% -*- mode: noweb; noweb-default-code-mode: R-mode; -*-
%\VignetteIndexEntry{nondetects - vignette}
%\VignetteDepends{nondetects, HTqPCR}
%\VignettePackage{nondetects}
\documentclass[12pt]{article}
\usepackage{hyperref, Sweave}
%\usepackage{natbib}

\textwidth=6.2in
\textheight=8.5in
\parskip=.3cm
\oddsidemargin=.1in
\evensidemargin=.1in
\headheight=-.3in

\newcommand\Rpackage[1]{{\textsf{#1}\index{#1 (package)}}}
\newcommand\dataset[1]{{\textit{#1}\index{#1 (data set)}}}
\newcommand\Rclass[1]{{\textit{#1}\index{#1 (class)}}}
\newcommand\Rfunction[1]{{{\small\texttt{#1}}\index{#1 (function)}}}
\newcommand\Rfunarg[1]{{\small\texttt{#1}}}
\newcommand\Robject[1]{{\small\texttt{#1}}}

\author{Matthew N. McCall}

\begin{document}
\title{Non-detects in qPCR data: methods to model and impute
  non-detects in the results of qPCR experiments (nondetects)}
\maketitle \tableofcontents

\section{Background on non-detects in qPCR data}
Quantitative real-time PCR (qPCR) measures gene expression for a
subset of genes through repeated cycles of sequence-specific DNA
amplification and expression measurements. During the exponential
amplification phase, each cycle results in an approximate doubling of
the quanitity of each target transcript. The threshold cycle (Ct) --
the cycle at which the target gene's expression first exceeds a
predetermined threshold -- is used to quantify the expression of each
target gene. These Ct values typically represent the raw data from a
qPCR experiment.

One challenge of qPCR data is the presence of \emph{non-detects} --
those reactions failing to attain the expression threshold. While most
current software replaces these non-detects with the maximum possible
Ct value (typically 40), recent work has shown that this introduces
large biases in estimation of both absolute and differential
expression. Here, we treat the non-detects as missing data, model the
missing data mechanism, and use this model to impute Ct values for the
non-detects.

\section{EM algorithm}
We propose the following model of observed expression for gene $i$,
sample-type $j$, and replicate $k$, $Y_{ijk}$:
\begin{displaymath}
Y_{ijk} = \left\{ \begin{array}{ll}
\theta_{ij} + \delta_{k} + \varepsilon_{ijk} & \textrm{if $Z_{ijk}=1$}\\
\textrm{non-detect} & \textrm{if $Z_{ijk}=0$}
\end{array} \right.
\end{displaymath}
where $\delta_{k}$ represents a global shift in expression across samples and,
\begin{displaymath}
Pr(Z_{ijk}=1) = \left\{ \begin{array}{ll}
g(Y_{ijk}) & \textrm{if $Y_{ijk} < 40$} \\
0 & \textrm{otherwise}
\end{array} \right.
\end{displaymath}
Here, $g(Y_{ijk})$ can be estimated via the following logistic regression:
\[
logit(Pr(Z_{ijk}=1)) =  \beta_0 + \beta_1 \hat{\theta}_{ij}
\]
where $\hat{\theta}_{ij}$ is an estimate of the average expression for
gene $i$ and sample-type $j$. 

\section{Example}

\subsection*{Data from Sampson \emph{et al.} Oncogene 2013}
Two cell types -- young adult mouse colon (YAMC) cells and
mutant-p53/activated-Ras transformed YAMC cells -- in combination with
three treatments -- untreated, sodium butyrate, or valproic acid. Four
replicates were performed for each cell-type/treatment combination
\cite{sampson2012gene}.

\subsection*{Load the data}
<<echo=T,results=hide>>=
library(HTqPCR)
library(nondetects)
data(oncogene2013)
@ 

\subsection*{Examine residuals when non-detects are replaced by 40}
Normalize to Becn1:
<<echo=T,results=hide>>=
normCt <- normalizeCtData(oncogene2013, norm = "deltaCt", 
                          deltaCt.genes = "Becn1")
@ 

Calculate residuals for each set of replicates:
<<echo=T,results=hide>>=
conds <- paste(pData(normCt)$sampleType,pData(normCt)$treatment,sep=":")
resids <- matrix(nrow=nrow(normCt), ncol=ncol(normCt))
for(i in 1:nrow(normCt)){
  for(j in 1:ncol(normCt)){
    ind <- which(conds==conds[j])
    resids[i,j] <- exprs(normCt)[i,j]-mean(exprs(normCt)[i,ind])
  }
}
@ 

Create boxplots of residuals stratified by the presence of a non-detect:
<<echo=T,results=hide>>=
iND <- which(featureCategory(normCt)=="Undetermined", arr.ind=TRUE)
iD <- which(featureCategory(normCt)!="Undetermined", arr.ind=TRUE)
boxes <- list("observed"=-resids[iD], "non-detect"=-resids[iND])
@ 
<<echo=T,fig=T>>=
boxplot(boxes, main="",ylim=c(-12,12),
        ylab=expression(paste("-",Delta,"Ct residuals",sep="")))
@ 

\subsection*{Impute non-detects}
<<echo=T,results=hide>>=
oncogene2013 <- qpcrImpute(oncogene2013, 
                           groupVars=c("sampleType","treatment"))
@ 

\subsection*{Examine residuals when non-detects are replaced by imputed values}
Normalize to Becn1:
<<echo=T,results=hide>>=
normCt <- normalizeCtData(oncogene2013, norm = "deltaCt", 
                          deltaCt.genes = "Becn1")
@ 

Remove the normalization gene:
<<echo=T,results=hide>>=
normCt <- normCt[-which(featureNames(normCt)=="Becn1"),]
@ 

Calculate residuals for each set of replicates:
<<echo=T,results=hide>>=
conds <- paste(pData(normCt)$sampleType,
               pData(normCt)$treatment,sep=":")
resids <- matrix(nrow=nrow(normCt), ncol=ncol(normCt))
for(i in 1:nrow(normCt)){
  for(j in 1:ncol(normCt)){
    ind <- which(conds==conds[j])
    resids[i,j] <- exprs(normCt)[i,j]-mean(exprs(normCt)[i,ind])
  }
}
@ 

Create boxplots of residuals stratified by the presence of a non-detect:
<<echo=T,results=hide>>=
iI <- which(featureCategory(normCt)=="Imputed", arr.ind=TRUE)
iD <- which(featureCategory(normCt)!="Imputed", arr.ind=TRUE)
boxes <- list("observed"=-resids[iD], "imputed"=-resids[iI])
@ 
<<echo=T,fig=T>>=
boxplot(boxes, main="",ylim=c(-12,12),
        ylab=expression(paste("-",Delta,"Ct residuals",sep="")))
@ 

\section{Additional examples}
Two additional example data sets are used in the paper and included in
the package. These are each briefly described below.

\subsection*{Data from Almudevar \emph{et al.} SAGMB 2011}
Cells transformed to malignancy by mutant p53 and activated Ras are
perturbed with the aim of restoring gene expression to levels found in
non-transformed parental cells via retrovirus-mediated re-expression
of corresponding cDNAs or shRNA-dependent stable knock-down. The data
contain 4-6 replicates for each perturbation, and each perturbation
has a corresponding control sample in which only the vector has been
added \cite{almudevar2011fitting}.

<<echo=T,results=hide>>=
library(nondetects)
data(sagmb2011)
@ 

\subsection*{Data from McMurray \emph{et al.} Nature 2008}
A study of the effect of p53 and/or Ras mutations on gene
expression. The third dataset is a comparison between four cell types
-- YAMC cells, mutant-p53 YAMC cells, activated-Ras YAMC cells, and
p53/Ras double mutant YAMC cells. Three replicates were performed for
the untransformed YAMC cells, and four replicates were performed for
each of the other cell types \cite{mcmurray2008synergistic}.

<<echo=T,results=hide>>=
library(nondetects)
data(nature2008)
@ 

\section*{Funding} 
This work was supported by National Institutes of Health [grant
  numbers CA009363, CA138249, HG006853]; and an Edelman-Gardner
Foundation Award.

\section{Session Info}
<<echo=T>>=
sessionInfo()
@ 

\bibliographystyle{abbrv}
\bibliography{nondetects}

\end{document}
